ui/pages/confirmation/confirmation.js

```diff
@@ -258,14 +262,28 @@ export default function ConfirmationPage({
     return INPUT_STATE_CONFIRMATIONS.includes(type);
   };
 
-  const handleSubmit = () =>
-    templateState[pendingConfirmation.id]?.useWarningModal
+  const handleSubmitResult = (submitResult) => {
+    if (submitResult?.length > 0) {
+      setLoadingText(templatedValues.submitText);
+      setSubmitAlerts(submitResult);
+      setLoading(true);
+    } else {
+      setLoading(false);
+    }
+  };
+  const handleSubmit = async () => {
+    setLoading(true);
+    // submit result is an array of errors or empty on success
+    const submitResult = await (templateState[pendingConfirmation.id]
+      ?.useWarningModal
       ? setShowWarningModal(true)
-      : templatedValues.onSubmit(
+      : await templatedValues.onSubmit(
           hasInputState(pendingConfirmation.type)
             ? inputStates[MESSAGE_TYPE.SNAP_DIALOG_PROMPT]
             : null,
-        );
+        ));
+    handleSubmitResult(submitResult);
```

Comment:
It seems that this is getting called even in the `setShowWarningModal` case. In that case, `submitResult` would be `undefined`, which results in `setLoading(false)` being called. This effectively breaks the loading state for confirmations with warnings (at least I think so - I didn't test this.

It might be simpler to change this ternary into if-else blocks, so that we can ensure this is only called when `onSubmit` is called.